testinfra/backend/kubectl.py

```diff
@@ -21,23 +21,41 @@ class KubectlBackend(base.BaseBackend):
     NAME = "kubectl"
 
     def __init__(self, name, *args, **kwargs):
-        self.name, self.user = self.parse_containerspec(name)
+        self.container = None
+        self.name, self.container = self.parse_podspec(name)
+
+        if kwargs:
+            if "namespace" in kwargs:
+                self.namespace = kwargs.get('namespace')
+            if "container" in kwargs:
+                self.container = kwargs.get('container')
+
         if "/" in self.name:
             self.name, self.container = self.name.split("/", 1)
-        else:
-            self.container = None
+
         super(KubectlBackend, self).__init__(self.name, *args, **kwargs)
 
     def run(self, command, *args, **kwargs):
         cmd = self.get_command(command, *args)
         # `kubectl exec` does not support specifying the user to run as.
         # See https://github.com/kubernetes/kubernetes/issues/30656
         if self.container is None:
-            out = self.run_local(
-                "kubectl exec %s -- /bin/sh -c %s", self.name, cmd)
+            if self.namespace:
+                out = self.run_local(
+                    "kubectl -n %s exec %s -- /bin/sh -c %s",
+                    self.namespace, self.name, cmd)
+            else:
+                out = self.run_local(
+                    "kubectl exec %s -- /bin/sh -c %s",
+                    self.name, cmd)
         else:
-            out = self.run_local(
-                "kubectl exec %s -c %s -- /bin/sh -c %s",
-                self.name, self.container, cmd)
+            if self.namespace:
+                out = self.run_local(
+                    "kubectl -n %s exec %s -c %s -- /bin/sh -c %s",
+                    self.namespace, self.name, self.container, cmd)
+            else:
+                out = self.run_local(
+                    "kubectl exec %s -c %s -- /bin/sh -c %s",
+                    self.name, self.container, cmd)
```

Comment:
Instead of handling the 4 cases we could build a `kubectlcmd` string and a `cmd_args` array.

````python
kubectlcmd = 'kubectl '
cmd_args = []
if self.namespace is not None:
    kubectlcmd += '-n %s '
    cmd_args.append(self.namespace)
if self.container is not None:
    kubectlcmd += '-c %s '
    cmd_args.append(self.container)
kubectlcmd += 'exec %s -- /bin/sh -c %s'
cmd_args.extend([self.name, cmd])
out = self.run_local(kubectlcmd, *cmd_args)
```